code-styling/bash/README.md

```diff
@@ -0,0 +1,423 @@
+
+- Functions MUST be defined without the `function` keyword, with round brackets, 
+  like this: `myfunc() { ... }`
+
+- All functions MUST be located together in the file just below constants. 
+  Executable code MUST NOT exists between functions.
+
+- Complex one-liners MUST be placed in standalone functions with an appropriate 
+  name and description.
+
+#### Function packaging
+
+- If it should be possible to call certain code both directly _and_ sourced from 
+  other scripts, it MUST be wrapped in a "self" check, like this: 
+  
+  ```bash
+  if [ "${BASH_SOURCE[0]}" == "${0}" ]; then
+      # direct call to file
+      main "${@}"
+  fi  # else file is included from another script
+  ```
+
+### Strings
+
+- For strings, single quotes `'` MUST be used when substitution is not required.
+- For strings, double quotes `"` MUST be used when substitution is required.
+
+### User messaging
+
+- User output SHOULD be formatted to follow the Heroku output format[^2](#references).
+- Warnings and errors MUST go to `STDERR`, anything than could be parsed SHOULD go to `STDOUT`.
+
+### Variables
+
+- Unless they are single character shell specials or arrays, variables MUST be 
+  "brace-quoted" (surrounded with curly braces `{` and `}`), 
+  like this: `${variable}`.  
+  Specials are `$!`, `$-`, `$_`, `$?`, `$#` `$*` `$@` `$$`.
+
+- Global variables SHOULD be avoided. When global variables _are_ used they MUST
+  be declared as `readonly` as much as possible.
+
+- Variables that are function-specific MUST be declared as `local`, unless `local` 
+  can not be used. (For instance when declaring as an associative array). For 
+  such cases `declare` MUST be used.
+ 
+- For the sake of clarity, read-only variables MUSt be declared using `readonly`.
+  Read-only variables MUST NOT be declared using `declare -r`.
+
+- Declaration and assignment SHOULD be placed on different lines.
+
+- To test variables that could be undeclared dot-slash `:-` MUST be used
+  (like this: `${variable:-}`)
+
+<!-- ======================================================================= -->
+
+## Common pitfalls
+
+Some more common problems and bugs in BASH can be avoided by adhering to the 
+following conventions:
+
+### Files and folders
+
+- Prefer absolute paths (leverage `$PWD`), always qualify relative paths with `./`.
+
+- When using `rm $directory/` make sure to add `:?` to ensure this never expands
+  to `rm /`, like this: `rm "${sPath:?}/${sTarget}"`
+
+- Use `mktemp` for temporary files, always cleanup with a `trap`.
+
+- Use `pushd /path/to/directory ` and `popd` (optionally redirected to `/dev/null`)
+  rather than `cd /path/to/directory` and `cd -`
+
+### Internal (or "builtin") commands
+
+- Shell builtin functions and BASH variable substitution SHOULD be used rather
+  than invoking a separate process (like `awk` or `sed`).
+
+- For complex string formatting `printf` SHOULD be used rather than `echo`.
+
+- the `eval` command SHOULD be avoided.
+
```

Comment:
Document exception for commands like ssh-agent which generate shell code as part of the operation